### ISSUE TITLE
[UX issue] fullscreen window + floating windows: do not behave as expected, and difficult to recover from

### ISSUE THREAD

**Issue by inga-lovinde** [riverwm/river#775]
For example:
* Open Firefox; open any page in it;
* Enter fullscreen mode with F11;
* Right-click, "save page as" (the same applies to "save link as", "save image as", etc; the same applies to chromium).

Expected result: floating file picker appears on top of a full-screen firefox view.
Actual result: nothing appears on the screen (presumably file picker is displayed behind firefox view), firefox no longer reacts to any key presses or clicks (presumably because the focus is on the file picker), F11 does not work too (presumably because the focus is on the file picker).

The same behavior applies to the other apps with full-screen views as well (e.g. in Telegram, images are opened on full screen, and trying to "save as" also leads to invisible but focused file picker).

Sometimes I can recover from this by repeatedly pressing Esc, apparently one of these key presses exits the invisible file picker and returns the focus to full-screen view.

Sometimes the only way to recover from this seems to be to hide the tag full-screen view belongs to (e.g. by Super+Number switching to another "desktop"), and killing the process of that view from the terminal.

I'm not sure if this is a bug or not, but this is certainly an UX issue; at least there should be a way to gracefully recover from this? And the best solution I think would be to display floating views on top of full-screen ones, not behind them.

**Comment by Leon-Plickat** [comment]
> (presumably because the focus is on the file picker)

river will never focus another window if there is a fullscreen window. This is firefox refusing to accept input on its primary toplevel if it also has a popup open. Try using a river keybind for exiting fullscreen.

> And the best solution I think would be to display floating views on top of full-screen ones, not behind them.

I doubt that would be good UX, as the common use case for fullscreen in river is to quickly and temporarily make only a single window visible, so you can focus on it. I personally tend to do so quite often and floating  windows appearing on top of fullscreened ones would be pretty annoying.

Also I am pretty sure it would be a protocol violation to display another toplevel on top of a fullscreen one.

Instead, I think we could make river automatically exit fullscreen if a new window spawn.

**Comment by inga-lovinde** [comment]
> This is firefox refusing to accept input on its primary toplevel if it also has a popup open

It's been a long time since I was involved with any desktop GUI programs, but I think that most, if not all, gtk/qt/etc programs cannot get any user input after opening a modal window, until modal window is closed?
And modal window here is gtk file picker, which is not quite managed by firefox.
So it's not like firefox is explicitly refusing to accept input; it's more that it (and any other applications) would have to go an extra mile to be able to accept input while modal window is open, if that's possible at all.

> Try using a river keybind for exiting fullscreen.

Thank you, WinF (with default river config), as opposed to F11, does indeed allow one to recover from this.

**Comment by adamcstephens** [comment]
> Instead, I think we could make river automatically exit fullscreen if a new window is mapped.

I think this would be an improvement. I've multiple times tried opening a vscode dialog and thinking I misclicked, only for there to be multiple dialogs open behind when I exit fullscreen.

**Comment by MaxVerevkin** [comment]
> > And the best solution I think would be to display floating views on top of full-screen ones, not behind them.
> 
> I doubt that would be good UX

I agree that displaying _all_ floating views above full-screen view is probably even worse that what we have now, but I think displaying _children_ floating top-levels above a full-screen parent might be what most users expect.

For example, I think firefox/GTK's modal file picker is a child view given this wayland debug message which appears when file picker is triggered:

```
[3456268.131]  -> xdg_toplevel@49.set_parent(xdg_toplevel@40)
```

And yes, it seems that displaying any floating view above full-screen one is against protocol docs.

> If the fullscreened surface is not opaque, the compositor must make sure that other screen content not part of the same surface tree (made up of subsurfaces, popups or similarly coupled surfaces) are not visible below the fullscreened surface.

**Comment by Leon-Plickat** [comment]
> So it's not like firefox is explicitly refusing to accept input;

I don't know Firefox or GTK internals, but at least protocol-wise is *trivial* to still get input on a toplevel that has a child view. In fact, you literally have to put in extra effort to not do so.

---

As for displaying child toplevels above fullscreened parents, I see the point, however the needed additional complexity in the focus and transaction system needed for that makes me prefer the simple approach of just automatically cancelling fullscreen.

**Comment by inga-lovinde** [comment]
> I don't know Firefox or GTK internals, but at least protocol-wise it is trivial to still get input on a toplevel that has a child view. In fact, you literally have to put in extra effort to not do so. So I personally would call it refusing.

I don't know GTK internals too, but:

1. [Wikipedia says](https://en.wikipedia.org/wiki/Modal_window): "_A modal window creates a [mode](https://en.wikipedia.org/wiki/Mode_(user_interface)) that disables the main window but keeps it visible, with the modal window as a child window in front of it. Users must interact with the modal window before they can return to the parent application._
2. I don't think I've ever seen a modal window (such as a file picker) _not_ blocking the main window, in any of the desktop GUI applications I've used, on Linux or on Windows.

There are also other interesting examples of modal window. E.g. open gedit, enter fullscreen with F11, enter some text without saving, try to exit with Super+Q (river's default) - now it's blocked (because there is an invisible "save changes before closing?" modal in background), and the only way to unblock is either to kill gedit from the terminal from another view, or to exit fullscreen with Super+F, which might be counterintuitive to users.

Not saying that this should be fixed (or how), but I think that's definitely a rough spot in UX. And when most, if not all, applications follow the same pattern with blocking modals, that's probably something we all have to accept, regardless of whether their developers are "refusing to accept input" or not?

**Comment by Leon-Plickat** [comment]
I'd argue that it's not our job to workaround bad design decisions in applications, even if they are popular. But that's actually irrelevant, because my proposed solution of automatically exiting fullscreen solves the input problem as well.

> There are also other interesting examples of modal window. E.g. open gedit, enter fullscreen with F11, enter some text without saving, try to exit with Super+Q (river's default) - now it's blocked (because there is an invisible "save changes before closing?" modal in background)

Here I'd argue that the "save changes?" dialog should have been a `xdg_popup` or a subsurface, not a child toplevel, but that is going off topic...

> > If the fullscreened surface is not opaque, the compositor must make sure that other screen content not part of the same surface tree (made up of subsurfaces, popups or similarly coupled surfaces) are not visible below the fullscreened surface.
>
> So basically the specs says that when one window is fullscreen, other windows' popups should not be displayed; but it does not say anything about this window's modals?

It actually does. What you call "modal" is in fact a separate toplevel with its own separate surface tree. The phrasing "similarly coupled surfaces" is a bit vague, but separate surface trees are explicitly excluded.

A surface tree for a toplevel consists of the main surface, an arbitrary amount of `wl_subsurface`s (commonly used to make your life a bit easier when rendering the UI) and an arbitrary amount of `xdg_popup`s (used for things like the context menu and some dialogs). These are all already rendered above fullscreen views, as otherwise the UI would be broken.

**Comment by inga-lovinde** [comment]
So, just to make sure, you're saying that this is not an issue with river, because all these applications open these modals in such a way that, according to wayland spec, these modals should not be visible in fullscreen mode, right?

And therefore it is either an issue with wayland spec (which does not explicitly contain any provisions for modals), or with frameworks such as gtk and qt, which implement wayland spec incorrectly by not using xdg_popup incorrectly?

I'm just trying to understand where to go from here, and where to file this bug. This seems like a huge UX issue but I was unable to find anything about it online, which is why I assumed it was related to river; but I guess people just don't use fullscreen in such a way with other, more widespread, compositors (or other compositors violate the spec in their fullscreen implementations)?

**Comment by inga-lovinde** [comment]
This is probably almost offtopic, but I've been looking at gedit sources and gtk docs, and gtk is more verbose than what I've worked with in the past (WinForms), and requires developers to explicitly set the dialog window as a modal: https://docs.gtk.org/gtk4/method.Window.set_modal.html

> Modal windows prevent interaction with other windows in the same application. To keep modal dialogs on top of main application windows, use gtk_window_set_transient_for() to make the dialog transient for the parent; most window managers will then disallow lowering the dialog below the parent.

also seemingly implying that dialogs are always first created as "a separate toplevel", if I understood you right?

At least that's [what gedit does](https://github.com/GNOME/gedit/blob/5f0d9deab0cfc48854e8c58dbcd2a1206c7bf9eb/gedit/gedit-close-confirmation-dialog.c): creates a new GeditCloseConfirmationDialog (which is a GtkMessageDialog which inherits GtkDialog which inherits GtkWindow), then calls `gtk_window_set_modal` to block the main window while the modal is open, and then calls `gtk_window_set_transient_for` "to make it transient for the parent".

According to GTK docs, "most window managers will then disallow lowering the dialog below the parent", but apparently according to wayland docs, lowering the dialog created in such a way below the parent is _mandatory_ in fullscreen mode? (Unless that GtkWindow is created as an xdg_popup)

So should I just close this issue and file an issue to GTK saying that GTK behavior (as documented, and as relied on by gedit) is different from the one described by wayland spec for fullscreen mode?

**Comment by ifreund** [comment]
I also find the spec unclear here, but I think rendering child toplevels over a fullscreen parent toplevel is not an unreasonable thing to do. I believe that's what sway and likely other compositors as well do here.

Perhaps automatically exiting fullscreen when a child toplevel is created is good enough though, it would certainly be simpler to implement and seems like it would be a step up over the current situation.

I don't think filing a GTK bug would get you anywhere, they're not going to change the behavior from how they want gtk to behave under GNOME, which is apparently with modal dialogs that disable input on the parent.

**Comment by inga-lovinde** [comment]
> I don't think filing a GTK bug would get you anywhere, they're not going to change the behavior

But maybe they will at least update the documentation to not say "most window managers will then disallow lowering the dialog below the parent" and mention that according to wayland spec, dialogs should still be lowered below the parent in fullscreen mode :)
(Assuming @Leon-Plickat 's interpretation of the spec is correct)

**Comment by novakne** [comment]
It seems that sway let the user choose

```
popup_during_fullscreen smart|ignore|leave_fullscreen
Determines what to do when a fullscreen view opens a dialog.
If smart (the default), the dialog will be displayed. If ignore, the
dialog will not be rendered. If leave_fullscreen, the view will exit
fullscreen mode and the dialog will be rendered.
```
